File: built-in plugins dsp/mb-delay-dub.cpp
/**
 * MB Dub Delay
 * Category : effect
 * Type     : delay
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic dub echo
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DELAY_DUB_H
#define MB_DELAY_DUB_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDelayDub : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-delay-dub";
    static constexpr const char* PLUGIN_NAME    = "MB Dub Delay";
    static constexpr const char* PLUGIN_TYPE    = "delay";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float time = 500f;  // range [100, 2000]
    float feedback = 0.6f;  // range [0, 0.95]
    float filter = 3000f;  // range [500, 10000]
    float mix = 0.4f;  // range [0, 1]
    };

    MbDelayDub() = default;
    ~MbDelayDub() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.time = std::clamp(params.time, 100f, 2000f);
        params.feedback = std::clamp(params.feedback, 0f, 0.95f);
        params.filter = std::clamp(params.filter, 500f, 10000f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Dub Delay
        return input;
    }
};

#endif // MB_DELAY_DUB_H

File: maxbooster7.5/built-in plugins dsp/mb-delay-dub.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbDelayDubPlugin: PluginDefinition = {
  id: "mb-delay-dub",
  slug: "mb-delay-dub",
  name: "MB Dub Delay",
  category: "effect",
  type: "delay",
  version: "1.0.0",
  description: "Classic dub echo",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "time",
      name: "Time",
      type: "float",
      defaultValue: 500,
      minValue: 100,
      maxValue: 2000,
      automatable: true,
    },
    {
      id: "feedback",
      name: "Feedback",
      type: "float",
      defaultValue: 0.6,
      minValue: 0,
      maxValue: 0.95,
      automatable: true,
    },
    {
      id: "filter",
      name: "Filter",
      type: "float",
      defaultValue: 3000,
      minValue: 500,
      maxValue: 10000,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { time: 500, feedback: 0.6, filter: 3000, mix: 0.4 },
};

export default MbDelayDubPlugin;
